catalog/rocks/catalog.go

```diff
@@ -0,0 +1,241 @@
+package rocks3
+
+import (
+	"context"
+	"time"
+)
+
+// Basic Types
+
+// Repository represents repository metadata
+type Repository struct {
+	StorageNamespace StorageNamespace
+	CreationDate     time.Time
+	DefaultBranch    BranchID
+}
+
+// Entry represents metadata or a given object (modified date, physical address, etc)
+type Entry struct {
+	LastModified time.Time
+	Address      string
+	Metadata     map[string]string
+	ETag         string
+}
+
+// Commit represents commit metadata (author, time, tree ID)
+type Commit struct {
+	Committer    string
+	Message      string
+	TreeID       TreeID
+	CreationDate time.Time
+	Parents      []CommitID
+	Metadata     map[string]string
+}
+
+// Branch is a pointer to a commit.
+type Branch struct {
+	CommitID CommitID
+	// nolint: structcheck, unused
+	stagingToken StagingToken
+}
+
+// Diff represents a changed state for a given entry (added, removed, changed, conflict)
+type DiffType uint8
+
+const (
+	DiffTypeAdded DiffType = iota
+	DiffTypeRemoved
+	DiffTypeChanged
+	DiffTypeConflict
+)
+
+type Diff struct {
+	Path Path
+	Type DiffType
+}
+
+// StorageNamespace is the URI to the storage location
```

Comment:
let's add a note about validation here (for example, that functions/methods receiving a StorageNamespace, RepositoryID etc as a type can assume it has passed validation and is a legal one.